File: 202.快乐数.cpp
/*
 * @lc app=leetcode.cn id=202 lang=cpp
 *
 * [202] 快乐数
 */

// @lc code=start
class Solution {
    public:
        bool isHappy(int n) {
            unordered_set<int> hash;
            while (n != 1 && !hash.count(n))
            {
                hash.insert(n);
                n = get_next(n);
            }
            return n == 1;
        }
    private:
        int get_next(int n)
        {
            int res = 0;
            while (n > 0)
            {
                int d = n % 10;
                res += d * d;
                n /= 10;
            }
            return res;
        }
    };
// @lc code=end
class Solution {
    public:
        bool isHappy(int n) {
            string s = to_string(n);
            set<string> hash;
            while (1)
            {
                int sums = 0;
                for (int i = 0; i < s.size(); i ++ )
                {
                    sums += (s[i] - '0') * (s[i] - '0');
                }
                if (sums == 1) return 1;
                sort(s.begin(), s.end());
                if (hash.count(s)) return 0;
                hash.insert(s);
                s = to_string(sums);
            }
            return 0;
        }
    };

File: 面经150/5.哈希表/202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#

# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        ss = set()
        while n != 1:
            tmp = str(n)
            x = 0
            for c in tmp:
                x += int(c) ** 2
            n = x
            if n in ss: return False
            else: ss.add(n)
        return True
# @lc code=end
class Solution:
    def isHappy(self, n: int) -> bool:
        ss = set()
        while n != 1:
            tmp = str(n)
            x = 0
            p = len(tmp) - 1
            cnt = [0] * 10
            for c in tmp:
                cnt[ord(c) - ord('0')] += 1
                x += int(c) ** 2
                p -= 1
            n = x
            if tuple(cnt) in ss: return False
            else: ss.add(tuple(cnt))
        return True
